src/mc_tools.jl

```diff
@@ -29,24 +26,24 @@ positive, and all rows must sum to unity
 """
 type MarkovChain{T<:Real}
     p::Matrix{T} # valid stochastic matrix
+end
 
-    function MarkovChain(p)
-        n, m = size(p)
+function MarkovChain{T<:Real}(p::Matrix{T})
+    n,m = size(p)
 
-        n != m && throw(ArgumentError("stochastic matrix must be square"))
-        any(p .< 0) &&
-            throw(ArgumentError("stochastic matrix must have nonnegative elements"))
-        isapprox(sum(p, 2), one(T)) ||
-            throw(ArgumentError("stochastic matrix rows must sum to 1"))
-        new(p)
+    if n != m 
+        throw(ArgumentError("stochastic matrix must be square"))
+    elseif any(p.<0)
+        throw(ArgumentError("stochastic matrix must have nonnegative elements"))
+    elseif any(x->!isapprox(sum(x),1), [p[i,:] for i in 1:n])
+        throw(ArgumentError("stochastic matrix rows must sum to 1"))
     end
-end
 
-# Provide constructor that infers T from eltype of matrix
-MarkovChain{T<:Real}(p::Matrix{T}) = MarkovChain{T}(p)
+    return MarkovChain{T}(p)
+end
```

Comment:
We should restore the inner+outer constructor pair. The reason is that we never want to be able to create an instance of `MarkovChain` without these 3 conditions holding -- these are the invariants referred to in the [inner constructor](http://julia.readthedocs.org/en/latest/manual/constructors/#inner-constructor-methods) section of the mainual